File: services/helm/openagent/files/mcp/prewarm.py
#!/usr/bin/env python3
"""Materialise the MCP package caches at boot WITHOUT starting any server.

The gateway's MCP client runs ``npx -y <pkg>`` / ``uvx <pkg>``, which installs
the package inside the *connect window*. Cancel that connect and the client
kills the child mid-reify; npm's install is not atomic, so the leftover
``node_modules`` looks complete and every later start skips the install and
dies on a missing module (observed: argocd ``Cannot find module
'pino-std-serializers'``, gistpad with no ``_npx`` at all).

The previous mechanism "fixed" this by *running the server* at boot and relying
on ``timeout`` to reap it. MCP servers do not exit on stdin EOF, and ``timeout``
only signals the direct child, so the ``node`` grandchildren leaked — verified
still alive 3h42m after boot, holding the servers' ports and memory.

This script replaces it:

* it resolves/installs each package with a command that never runs the server
  (``npm exec --package=<pkg> -- true`` for npm, ``uv tool install <spec>``
  for uv);
* every child runs in its own session/process group and is hard-killed
  (SIGTERM then SIGKILL to the *group*) on timeout, so nothing can outlive it;
* it sweeps ``_npx`` / npm-cache orphans left behind by older boots (they are
  reparented to PID 1, which is how we recognise them) and asserts none remain;
* it reads the package list from the rendered manifest, so there is no second
  copy of the pins to drift from values.yaml.

Booted as root, before the gateway starts; non-fatal on failure.
"""

from __future__ import annotations

import json
import os
import re
import signal
import subprocess
import sys
import time
from concurrent.futures import ThreadPoolExecutor
from typing import Any

DEFAULT_MANIFEST = "/opt/data/mcp-verify/mcp-manifest"
NPM_TIMEOUT = 300
UV_TIMEOUT = 600
NPM_PARALLELISM = 4


def log(message: str) -> None:
    print(f"[mcp-prewarm] {message}", flush=True)


def load_servers(path: str) -> dict[str, Any]:
    with open(path, encoding="utf-8") as handle:
        return json.load(handle)


def package_spec(server: dict[str, Any]) -> tuple[str, str] | None:
    """Return (kind, spec) for an npx/uvx server, else None.

    Handles ``npx``/``uvx`` directly and shell-wrapped forms
    (``sh -c "exec npx -y pkg | grep jsonrpc"``).
    """
    command = str(server.get("command", ""))
    parts = [command] + [str(a) for a in server.get("args", []) or []]
    tokens = re.split(r"[\s;|&()]+", " ".join(parts))
    for index, token in enumerate(tokens):
        if token not in ("npx", "uvx"):
            continue
        kind = "npm" if token == "npx" else "uv"
        for candidate in tokens[index + 1 :]:
            if not candidate or candidate.startswith("-") or candidate in ("exec", "run", "dlx"):
                continue
            return kind, candidate
    return None


def run_bounded(cmd: list[str], env: dict[str, str], timeout: int) -> tuple[int, str]:
    """Run *cmd* in its own process group; SIGKILL the group on timeout."""
    proc = subprocess.Popen(
        cmd,
        stdin=subprocess.DEVNULL,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.PIPE,
        env=env,
        start_new_session=True,
    )
    try:
        _, err = proc.communicate(timeout=timeout)
        return proc.returncode, (err or b"").decode("utf-8", "replace")
    except subprocess.TimeoutExpired:
        try:
            os.killpg(os.getpgid(proc.pid), signal.SIGTERM)
            time.sleep(2)
            os.killpg(os.getpgid(proc.pid), signal.SIGKILL)
        except ProcessLookupError:
            pass
        proc.communicate()
        return -1, f"timed out after {timeout}s (process group killed)"


def npm_cache_for(server: dict[str, Any]) -> str:
    env = server.get("env") or {}
    return (
        env.get("npm_config_cache")
        or os.environ.get("NPM_CONFIG_CACHE")
        or os.path.join(os.environ.get("HOME", "/root"), ".npm")
    )


def materialise_npm(name: str, spec: str, cache: str, env: dict[str, str]) -> bool:
    marker = os.path.join(cache, ".mcp-prewarm")
    if os.path.exists(marker) and os.path.isdir(os.path.join(cache, "_npx")):
        try:
            with open(marker, encoding="utf-8") as handle:
                if handle.read().strip() == spec:
                    log(f"{name}: cache already warm ({spec})")
                    return True
        except OSError:
            pass
    # A cache without a matching marker is either cold or poisoned (the
    # original failure mode). Drop _npx so npm reifies from scratch.
    _rmtree(os.path.join(cache, "_npx"))
    os.makedirs(cache, exist_ok=True)
    child_env = dict(env)
    child_env["NPM_CONFIG_CACHE"] = cache
    rc, err = run_bounded(
        ["npm", "exec", "--yes", f"--package={spec}", "--", "true"],
        child_env,
        NPM_TIMEOUT,
    )
    if rc == 0:
        with open(marker, "w", encoding="utf-8") as handle:
            handle.write(spec)
        log(f"{name}: materialised {spec} into {cache}")
        return True
    log(f"{name}: FAILED to materialise {spec} (rc={rc}) {err.strip()[:400]}")
    return False


def materialise_uv(name: str, spec: str, env: dict[str, str]) -> bool:
    rc, err = run_bounded(["uv", "tool", "install", "--quiet", spec], env, UV_TIMEOUT)
    if rc == 0:
        log(f"{name}: materialised {spec} (uv tool)")
        return True
    log(f"{name}: FAILED to materialise {spec} (rc={rc}) {err.strip()[:400]}")
    return False


def _rmtree(path: str) -> None:
    import shutil

    shutil.rmtree(path, ignore_errors=True)


def _cmdline(pid: int) -> str:
    try:
        with open(f"/proc/{pid}/cmdline", "rb") as handle:
            return handle.read().decode("utf-8", "replace").replace("\0", " ").strip()
    except OSError:
        return ""


def _ppid(pid: int) -> int:
    try:
        with open(f"/proc/{pid}/stat", encoding="utf-8") as handle:
            return int(handle.read().rsplit(")", 1)[1].split()[1])
    except (OSError, IndexError, ValueError):
        return 0


def _looks_like_prewarm_orphan(cmd: str) -> bool:
    """True only for a node/npm process rooted in an npm MCP cache.

    Matching a bare package name would kill unrelated processes that merely
    mention it (e.g. an operator's shell), so require the cache path *and* a
    node-family executable or a node_modules path. Package pre-warm leftovers
    are exactly ``node .../<cache>/_npx/<hash>/node_modules/...``.
    """
    if "_npx" not in cmd and ".npm-mcp" not in cmd:
        return False
    argv0 = os.path.basename(cmd.split(" ", 1)[0])
    return argv0 in ("node", "npm", "npx") or "node_modules" in cmd


def sweep_orphans() -> list[tuple[int, str]]:
    """Kill reparented (ppid==1) leftovers from older pre-warm boots."""
    killed: list[tuple[int, str]] = []
    for entry in os.listdir("/proc"):
        if not entry.isdigit():
            continue
        pid = int(entry)
        if pid == os.getpid():
            continue
        cmd = _cmdline(pid)
        if not cmd or _ppid(pid) != 1:
            continue
        if _looks_like_prewarm_orphan(cmd):
            try:
                os.kill(pid, signal.SIGKILL)
                killed.append((pid, cmd[:160]))
            except OSError:
                pass
    return killed


def main() -> int:
    manifest = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_MANIFEST
    try:
        servers = load_servers(manifest)
    except OSError as exc:
        log(f"manifest {manifest} unavailable ({exc}); skipping pre-warm")
        return 0

    npm_specs: list[tuple[str, str, str]] = []
    uv_specs: list[tuple[str, str]] = []
    for name, server in servers.items():
        found = package_spec(server or {})
        if not found:
            continue
        kind, spec = found
        if kind == "npm":
            npm_specs.append((name, spec, npm_cache_for(server)))
        else:
            uv_specs.append((name, spec))

    # Sweep BEFORE materialising so we never touch our own children.
    killed = sweep_orphans()
    for pid, cmd in killed:
        log(f"killed orphan pid {pid}: {cmd}")
    log(f"swept {len(killed)} orphaned pre-warm process(es)")

    env = dict(os.environ)
    failures = 0
    if npm_specs:
        with ThreadPoolExecutor(max_workers=NPM_PARALLELISM) as pool:
            results = list(
                pool.map(lambda item: materialise_npm(*item, env), npm_specs)
            )
        failures += sum(1 for ok in results if not ok)
    for name, spec in uv_specs:
        if not materialise_uv(name, spec, env):
            failures += 1

    remaining = sweep_orphans()
    log(f"{len(remaining)} pre-warm orphan(s) remain after materialisation")
    if failures:
        log(f"{failures} package(s) failed to materialise (non-fatal)")
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: .useful-scripts/gitopsctl/internal/mcp/prewarm.go
package mcp

import (
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"syscall"
	"time"
)

const (
	prewarmMarker    = ".mcp-prewarm"
	prewarmNPMTimout = 300 * time.Second
	prewarmUVTimeout = 600 * time.Second
)

func prewarmLog(format string, args ...any) {
	fmt.Printf("[mcp-prewarm] %s\n", fmt.Sprintf(format, args...))
}

// PackageSpec returns (kind, spec) for an npx/uvx server, else ("", ""),
// handling shell-wrapped forms (sh -c "exec npx -y pkg | grep jsonrpc").
func PackageSpec(server Server) (string, string) {
	parts := []string{server.Command}
	parts = append(parts, server.Args...)
	tokens := tokenRe.Split(strings.Join(parts, " "), -1)
	for i, token := range tokens {
		if token != "npx" && token != "uvx" {
			continue
		}
		kind := "npm"
		if token == "uvx" {
			kind = "uv"
		}
		for _, candidate := range tokens[i+1:] {
			if candidate == "" || strings.HasPrefix(candidate, "-") || candidate == "exec" || candidate == "run" || candidate == "dlx" {
				continue
			}
			return kind, candidate
		}
	}
	return "", ""
}

var tokenRe = regexp.MustCompile(`[\s;|&()]+`)

// runBounded runs cmd in its own session/process group and SIGKILLs the group
// on timeout, so nothing can outlive the bound.
func runBounded(cmd *exec.Cmd, timeout time.Duration) (int, string) {
	cmd.Stdin = nil
	var errBuf bytes.Buffer
	cmd.Stderr = &errBuf
	cmd.SysProcAttr = &syscall.SysProcAttr{Setsid: true}
	if err := cmd.Start(); err != nil {
		return -1, err.Error()
	}
	done := make(chan error, 1)
	go func() { done <- cmd.Wait() }()
	select {
	case <-done:
		return cmd.ProcessState.ExitCode(), errBuf.String()
	case <-time.After(timeout):
		pgid := cmd.Process.Pid
		_ = syscall.Kill(-pgid, syscall.SIGTERM)
		time.Sleep(2 * time.Second)
		_ = syscall.Kill(-pgid, syscall.SIGKILL)
		<-done
		return -1, fmt.Sprintf("timed out after %s (process group killed)", timeout)
	}
}

func materialiseNPM(name, spec, cache string, env []string) bool {
	marker := filepath.Join(cache, prewarmMarker)
	if _, err := os.Stat(marker); err == nil {
		if data, err := os.ReadFile(marker); err == nil && strings.TrimSpace(string(data)) == spec {
			if info, err := os.Stat(filepath.Join(cache, "_npx")); err == nil && info.IsDir() {
				prewarmLog("%s: cache already warm (%s)", name, spec)
				return true
			}
		}
	}
	_ = os.RemoveAll(filepath.Join(cache, "_npx"))
	if err := os.MkdirAll(cache, 0o755); err != nil {
		prewarmLog("%s: cannot create cache %s: %v", name, cache, err)
		return false
	}
	cmd := exec.Command("npm", "exec", "--yes", "--package="+spec, "--", "true")
	cmd.Env = append(env, "NPM_CONFIG_CACHE="+cache)
	rc, errOut := runBounded(cmd, prewarmNPMTimout)
	if rc == 0 {
		if err := os.WriteFile(marker, []byte(spec), 0o644); err != nil {
			prewarmLog("%s: cannot write marker: %v", name, err)
			return false
		}
		prewarmLog("%s: materialised %s into %s", name, spec, cache)
		return true
	}
	prewarmLog("%s: FAILED to materialise %s (rc=%d) %s", name, spec, rc, snippet(errOut))
	return false
}

func materialiseUV(name, spec string, env []string) bool {
	cmd := exec.Command("uv", "tool", "install", "--quiet", spec)
	cmd.Env = env
	rc, errOut := runBounded(cmd, prewarmUVTimeout)
	if rc == 0 {
		prewarmLog("%s: materialised %s (uv tool)", name, spec)
		return true
	}
	prewarmLog("%s: FAILED to materialise %s (rc=%d) %s", name, spec, rc, snippet(errOut))
	return false
}

func cmdline(pid int) string {
	data, err := os.ReadFile(fmt.Sprintf("/proc/%d/cmdline", pid))
	if err != nil {
		return ""
	}
	return strings.TrimSpace(strings.ReplaceAll(string(data), "\x00", " "))
}

func ppid(pid int) int {
	data, err := os.ReadFile(fmt.Sprintf("/proc/%d/stat", pid))
	if err != nil {
		return 0
	}
	after := data
	if idx := bytes.LastIndexByte(after, ')'); idx >= 0 {
		fields := strings.Fields(string(after[idx+1:]))
		if len(fields) > 1 {
			if n, err := strconv.Atoi(fields[1]); err == nil {
				return n
			}
		}
	}
	return 0
}

func looksLikePrewarmOrphan(cmd string) bool {
	if !strings.Contains(cmd, "_npx") && !strings.Contains(cmd, ".npm-mcp") {
		return false
	}
	argv0 := filepath.Base(strings.SplitN(cmd, " ", 2)[0])
	return argv0 == "node" || argv0 == "npm" || argv0 == "npx" || strings.Contains(cmd, "node_modules")
}

// sweepOrphans kills reparented (ppid==1) leftovers from older pre-warm boots.
func sweepOrphans() [][2]string {
	var killed [][2]string
	entries, err := os.ReadDir("/proc")
	if err != nil {
		return killed
	}
	for _, entry := range entries {
		pid, err := strconv.Atoi(entry.Name())
		if err != nil || pid == os.Getpid() {
			continue
		}
		cmd := cmdline(pid)
		if cmd == "" || ppid(pid) != 1 {
			continue
		}
		if looksLikePrewarmOrphan(cmd) {
			if err := syscall.Kill(pid, syscall.SIGKILL); err == nil {
				if len(cmd) > 160 {
					cmd = cmd[:160]
				}
				killed = append(killed, [2]string{strconv.Itoa(pid), cmd})
			}
		}
	}
	return killed
}

// Prewarm materialises the npx/uvx package caches at boot without starting any
// server. Returns 1 when packages failed to materialise (non-fatal at boot).
func Prewarm(manifestPath string, parallel int) int {
	servers, err := LoadManifest(manifestPath)
	if err != nil {
		prewarmLog("manifest %s unavailable (%v); skipping pre-warm", manifestPath, err)
		return 0
	}

	type npmJob struct{ name, spec, cache string }
	var npmJobs []npmJob
	var uvJobs [][2]string
	for name, server := range servers {
		if server == nil {
			continue
		}
		kind, spec := PackageSpec(*server)
		if kind == "" {
			continue
		}
		if kind == "npm" {
			npmJobs = append(npmJobs, npmJob{name, spec, NPMCacheFor(*server)})
		} else {
			uvJobs = append(uvJobs, [2]string{name, spec})
		}
	}

	// Sweep BEFORE materialising so we never touch our own children.
	killed := sweepOrphans()
	for _, k := range killed {
		prewarmLog("killed orphan pid %s: %s", k[0], k[1])
	}
	prewarmLog("swept %d orphaned pre-warm process(es)", len(killed))

	env := os.Environ()
	failures := 0
	done := make(chan bool, len(npmJobs))
	sem := make(chan struct{}, parallel)
	for _, job := range npmJobs {
		go func(job npmJob) {
			sem <- struct{}{}
			defer func() { <-sem }()
			done <- materialiseNPM(job.name, job.spec, job.cache, env)
		}(job)
	}
	for range npmJobs {
		if !<-done {
			failures++
		}
	}
	for _, job := range uvJobs {
		if !materialiseUV(job[0], job[1], env) {
			failures++
		}
	}

	remaining := sweepOrphans()
	prewarmLog("%d pre-warm orphan(s) remain after materialisation", len(remaining))
	if failures > 0 {
		prewarmLog("%d package(s) failed to materialise (non-fatal)", failures)
		return 1
	}
	return 0
}
